Import Libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Display all columns
pd.set_option('display.max_columns', None)

Read The Excel Dataset file

In [4]:
df = pd.read_excel(
    "../data/raw/online_retail_II.xlsx",
    sheet_name="Year 2010-2011"
)

Preview the Dataset

In [5]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


Dataset Dimensions

In [6]:
df.shape

(541910, 8)

Column Names

In [7]:
df.columns

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country'],
      dtype='str')

Dataset Information

In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      541910 non-null  object        
 1   StockCode    541910 non-null  object        
 2   Description  540456 non-null  object        
 3   Quantity     541910 non-null  int64         
 4   InvoiceDate  541910 non-null  datetime64[us]
 5   Price        541910 non-null  float64       
 6   Customer ID  406830 non-null  float64       
 7   Country      541910 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


### Observations

- The dataset contains **541,910 transaction records** and **8 columns**.
- Most columns have complete data.
- The **Description** column contains some missing values.
- The **Customer ID** column has a significant number of missing values.
- The **InvoiceDate** column has already been recognized as a datetime data type, which is useful for time-based analysis.
- The **Quantity** column is stored as an integer and **Price** as a floating-point number, making them suitable for numerical analysis.
- The **Country** column is stored as a string, while **Invoice** and **StockCode** are stored as object data types.

Stastical summary

In [16]:
df.describe()


,Quantity,InvoiceDate,Price,Customer ID
count,541910.000000,541910,541910.000000,406830.000000
mean,9.552234,2011-07-04 13:35:22.342307,4.611138,15287.684160
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.080957,NaN,96.759765,1713.603074


In [17]:
df.describe(include=["object", "string"])

,Invoice,StockCode,Description,Country
count,541910,541910,540456,541910
unique,25900,4070,4223,38
top,573585,85123A,WHITE HANGING HEART T-LIGHT HOLDER,United Kingdom
freq,1114,2313,2369,495478


In [18]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    135080
Country             0
dtype: int64

# Observations

- The dataset contains missing values in two columns:
  - **Description:** 1,454 missing values.
  - **Customer ID:** 135,080 missing values.
- All remaining columns contain complete data.
- The large number of missing Customer IDs is significant because customer identification is essential for customer segmentation.
- Missing descriptions are relatively few and will be investigated during the data cleaning phase.

Check for Duplicate Rows

In [15]:
df.duplicated().sum()

np.int64(5268)

### Observations

- The dataset contains **5,268 duplicate records**.
- Duplicate transactions may arise due to repeated data entries or duplicate invoice records.
- At this stage, we will only identify their presence.
- Whether these duplicates should be removed will be determined during the data cleaning phase after further investigation.

# Conclusion

The Online Retail II dataset contains over **541,000** transaction records with **8 features** describing invoices, products, customers, and purchase information.

The initial inspection revealed:

- Missing values in the **Description** and **Customer ID** columns.
- A considerable number of transactions without customer identification.
- **5,268 duplicate records** that require further investigation.
- Appropriate data types for numerical and datetime features.

The dataset appears suitable for customer segmentation, but preprocessing and data cleaning are required before feature engineering and machine learning.